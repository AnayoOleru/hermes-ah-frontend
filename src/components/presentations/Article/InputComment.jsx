import React from 'react';
import PropTypes from 'prop-types';

const InputComment = ({
  id,
  commentVal,
  imageUrl,
  btnValue,
  submitForm,
  handleChange,
  enterKeyFormSubmit,
  handleClose,
  closeVal,
  placeholderValue,
}) => {
  return (
    <div className="write-comment edit-comment-margin" key={id}>
      <form className="comment-text-wrap" id={id} onSubmit={submitForm}>
        <div className="wrap-img-text">
          <div className="image">
            <img
              src={
                !imageUrl
                  ? 'https://res.cloudinary.com/sojidan/image/upload/v1557149927/avatar.png'
                  : imageUrl
              }
              alt="author"
            />
          </div>
          <textarea
            required
            type="text"
            id={id}
            name="comment-text"
            placeholder={placeholderValue}
            className="text-area-comm"
            onKeyDown={enterKeyFormSubmit}
            onChange={handleChange}
            value={commentVal}
          />
        </div>
        <div className="sub-form-cont">
          <button type="submit" className="submit-comment">
            {btnValue}
          </button>
          <button type="button" className="close" onClick={handleClose}>
            {closeVal}
          </button>
        </div>
        <div className="wrap-guide">
          <div className="input-guide">
            <span>Press </span>
            <img
              src="https://res.cloudinary.com/mchardex/image/upload/v1559145799/keyboard-key-shift-512.png"
              alt="shift"
            />
            <span> + </span>
            <img
              src="https://res.cloudinary.com/mchardex/image/upload/v1559145991/keyboard-key-enter-512.png"
              alt="enter"
            />
            <span>to move to next line</span>
          </div>
          <span className="space">|</span>
          <div className="input-guide-enter">
            <span>Press </span>
            <img
              src="https://res.cloudinary.com/mchardex/image/upload/v1559145991/keyboard-key-enter-512.png"
              alt="enter"
            />
            <span>to submit</span>
          </div>
        </div>
      </form>
    </div>
  );
};

InputComment.defaultProps = {
  imageUrl:
    'https://res.cloudinary.com/sojidan/image/upload/v1557149927/avatar.png',
  id: '',
  commentVal: '',
  closeVal: '',
  placeholderValue: '',
  btnValue: '',
  handleClose: () => {},
  submitForm: () => {},
  enterKeyFormSubmit: () => {},
  handleChange: () => {},
};

InputComment.propTypes = {
  imageUrl: PropTypes.string,
  closeVal: PropTypes.string,
  id: PropTypes.string,
  btnValue: PropTypes.string,
  placeholderValue: PropTypes.string,
  submitForm: PropTypes.func,
  commentVal: PropTypes.string,
  handleChange: PropTypes.func,
  enterKeyFormSubmit: PropTypes.func,
  handleClose: PropTypes.func,
};

export default InputComment;
