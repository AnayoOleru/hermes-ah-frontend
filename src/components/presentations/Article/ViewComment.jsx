import React, { Component } from 'react';
import TimeAgo from 'react-timeago';
import PropTypes from 'prop-types';
import InputComment from './InputComment';
import ViewEditHistoryModal from '../../shared/Modals/Modal';
import DeleteModal from '../../shared/Modals/DeleteModal';

class ViewComment extends Component {
  constructor(props) {
    super(props);
    this.state = {
      toggle: false,
      input: false,
      menu: false,
      showEdit: false,
      replyVal: '',
      showEditHistory: false,
      showDelete: false,
    };
  }

  toggleReplies = () => {
    const { toggle } = this.state;
    this.setState({ toggle: !toggle });
  };

  toggleCommentMenu = () => {
    const { menu } = this.state;
    this.setState({ menu: !menu });
  };

  showCommentInput = () => {
    const { input } = this.state;
    this.setState({ input: !input });
  };

  showEditComment = () => {
    const { showEdit } = this.state;
    this.setState({ showEdit: !showEdit, menu: false });
  };

  showDeleteModal = () => {
    this.setState({ showDelete: true });
  };

  closeDeleteModal = () => {
    this.setState({ showDelete: false, menu: false });
  };

  handleReplyInput = e => {
    const { value } = e.target;
    this.setState({ replyVal: value });
  };

  showEditHistoryModal = async e => {
    const { getCommentHistory } = this.props;
    const commentId = e.target.id;
    await getCommentHistory(commentId);
    this.setState({ showEditHistory: true, menu: false });
  };

  closeEditHistoryModal = () => {
    this.setState({ showEditHistory: false });
  };

  editComment = e => {
    e.preventDefault();
    const { replyVal } = this.state;
    const data = {
      body: replyVal,
    };
    const commentId = e.target.id;
    this.setState({ replyVal: '' });
    const { updateComment } = this.props;
    updateComment(data, commentId);
    this.showEditComment();
  };

  onEnterSubmitEdit = e => {
    if (e.keyCode === 13 && e.shiftKey === false) {
      e.preventDefault();
      const { replyVal } = this.state;
      const data = {
        body: replyVal,
      };
      const commentId = e.target.id;
      this.setState({ replyVal: '' });
      const { updateComment } = this.props;
      updateComment(data, commentId);
      this.showEditComment();
    }
  };

  deleteUserComment = e => {
    const commentId = e.target.id;
    const { deleteComment } = this.props;
    deleteComment(commentId);
  };

  render() {
    const {
      toggle,
      input,
      menu,
      showEdit,
      replyVal,
      showEditHistory,
      showDelete,
    } = this.state;
    const { comment, commentHistory, profile, likeComment } = this.props;
    return (
      <React.Fragment>
        {comment ? (
          <div className="comment-cont">
            {showEdit ? (
              <InputComment
                imageUrl={comment && comment.commentOwner.image_url}
                id={comment && comment.id}
                btnValue="Update Comment"
                closeVal="Cancel"
                commentVal={replyVal}
                placeholderValue={comment && comment.body}
                handleClose={this.showEditComment}
                handleChange={this.handleReplyInput}
                submitForm={this.editComment}
                enterKeyFormSubmit={this.onEnterSubmitEdit}
              />
            ) : (
              <div>
                <div className="comments-cont-text-content">
                  <div className="wrap-image-author">
                    {comment.commentOwner && (
                      <div className="image-author">
                        <div className="commenter-img">
                          <img
                            src={
                              !comment.commentOwner.image_url
                                ? 'https://res.cloudinary.com/sojidan/image/upload/v1557149927/avatar.png'
                                : comment.commentOwner.image_url
                            }
                            alt="author"
                          />
                        </div>
                        <div className="commenter-details">
                          <h3>
                            {`${comment.commentOwner.first_name} ${
                              comment.commentOwner.last_name
                            }`}
                          </h3>
                          <span>
                            {new Date(comment.updatedAt).toDateString()}
                          </span>
                        </div>
                      </div>
                    )}
                    <div className="toggle-comment-menu">
                      {profile && profile.id === comment.user_id ? (
                        <button type="button" onClick={this.toggleCommentMenu}>
                          <i className="fas fa-ellipsis-v" />
                        </button>
                      ) : null}
                      {menu && (
                        <div className="comment-menu">
                          <button type="button" onClick={this.showEditComment}>
                            Edit comment
                          </button>
                          <button type="button" onClick={this.showDeleteModal}>
                            Delete Comment
                          </button>
                        </div>
                      )}
                      <DeleteModal
                        open={showDelete}
                        closeConfirmationModal={this.closeDeleteModal}
                        id={comment.id}
                        buttonEvent={this.deleteUserComment}
                      />
                    </div>
                  </div>
                  {commentHistory &&
                    commentHistory.map(history => (
                      <ViewEditHistoryModal
                        key={history.id}
                        modalOpen={showEditHistory}
                        closeModal={this.closeEditHistoryModal}
                        title="Edit History"
                      >
                        {/* histories */}
                        {history.histories.map(histories => (
                          <div className="history-body" key={histories.id}>
                            <img
                              alt="comment-owner"
                              src={profile && profile.image_url}
                            />
                            <div className="author-details-wrap">
                              <div className="author">
                                <p className="auth-name">
                                  {`${profile.first_name} ${profile.last_name}`}
                                </p>
                                <p className="comment">{histories.body}</p>
                              </div>
                              <div className="time">
                                <p>
                                  <TimeAgo
                                    date={new Date(
                                      histories.createdAt
                                    ).toUTCString()}
                                  />
                                </p>
                              </div>
                            </div>
                            {/* histories */}
                          </div>
                        ))}
                        <div className="history-body">
                          <img
                            alt="comment-owner"
                            src={profile && profile.image_url}
                          />
                          <div className="author-details-wrap">
                            <div className="author">
                              <p className="auth-name">
                                {`${profile.first_name} ${profile.last_name}`}
                              </p>
                              <p className="comment">{history.body}</p>
                            </div>
                            <div className="time">
                              <p>
                                <TimeAgo
                                  date={new Date(
                                    history.updatedAt
                                  ).toUTCString()}
                                />
                              </p>
                            </div>
                          </div>
                        </div>
                      </ViewEditHistoryModal>
                    ))}
                  <div className="commenter-text">
                    <p>{comment.body}</p>
                  </div>
                  <div className="commt-details">
                    <p className="last-upd">
                      <TimeAgo
                        date={new Date(comment.createdAt).toUTCString()}
                      />
                    </p>
                    <span>.</span>
                    <button
                      type="button"
                      className="reply"
                      onClick={this.showCommentInput}
                    >
                      Reply
                    </button>
                    <div className="like-comment">
                      <button
                        type="button"
                        onClick={() => likeComment(comment.id)}
                      >
                        <i className="far fa-thumbs-up" />
                      </button>
                      <span>{comment.likes_count}</span>
                    </div>
                    {profile.id === comment.user_id &&
                    new Date(comment.updatedAt).toUTCString() >
                      new Date(comment.createdAt).toUTCString() ? (
                      // eslint-disable-next-line react/jsx-indent
                      <div>
                        <span>.</span>
                        <span
                          className="edited-comm"
                          id={comment && comment.id}
                          onClick={this.showEditHistoryModal}
                          onKeyDown={this.showEditHistoryModal}
                          role="presentation"
                        >
                          Edited
                        </span>
                      </div>
                    ) : null}
                  </div>
                  {input ? <InputComment /> : null}
                  {comment.replies && comment.replies.length > 0 ? (
                    <div className="hide-replies">
                      <hr className="line" />
                      <button type="button" onClick={this.toggleReplies}>
                        {toggle ? 'Hide replies' : 'Show replies'}
                      </button>
                    </div>
                  ) : null}
                  {toggle ? (
                    <div>
                      {comment.replies.map(reply => (
                        <div className="response" key={reply.createdAt}>
                          <div className="wrap-image-author">
                            <div className="image-author">
                              <div className="commenter-img">
                                <img
                                  src={reply.replier.image_url}
                                  alt="author"
                                />
                              </div>
                              <div className="commenter-details">
                                <h3>
                                  {`${reply.replier.first_name} ${
                                    reply.replier.last_name
                                  }`}
                                </h3>
                                <span>
                                  {new Date(reply.createdAt).toDateString()}
                                </span>
                              </div>
                            </div>
                            <i className="fas fa-ellipsis-v" />
                          </div>
                          <div className="commenter-text">
                            <p>{reply.reply}</p>
                          </div>
                          <div className="commt-details">
                            <p className="last-upd">
                              {
                                <TimeAgo
                                  date={new Date(reply.createdAt).toUTCString()}
                                />
                              }
                            </p>
                          </div>
                        </div>
                      ))}
                    </div>
                  ) : null}
                </div>
              </div>
            )}
          </div>
        ) : null}
      </React.Fragment>
    );
  }
}

ViewComment.defaultProps = {
  profile: {},
  commentHistory: [{}],
};

ViewComment.propTypes = {
  comment: PropTypes.shape().isRequired,
  profile: PropTypes.shape(),
  commentHistory: PropTypes.arrayOf(PropTypes.shape()),
  updateComment: PropTypes.func.isRequired,
  getCommentHistory: PropTypes.func.isRequired,
  likeComment: PropTypes.func.isRequired,
  deleteComment: PropTypes.func.isRequired,
};

export default ViewComment;
