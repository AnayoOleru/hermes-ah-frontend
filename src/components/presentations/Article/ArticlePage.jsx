/* eslint-disable react/jsx-no-target-blank */
import React, { Component } from 'react';
import { Grid } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import NavBar from '../../shared/NavBar/NavBar';
import ViewComment from './ViewComment';
import ReadingCard from './ReadingArticleCard';
import InputComment from './InputComment';
import Loader from '../../shared/Loader/Loader';
import './Article.scss';
import Rate from './Rate';

class ArticlePage extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  async componentDidMount() {
    const { getSingleArticle, match } = this.props;
    const { articleId } = match.params;
    await getSingleArticle(articleId);
  }

  render() {
    const {
      singleArticle,
      match,
      postComment,
      rateArticle,
      isLoadingReducer,
      reportArticle,
      user,
    } = this.props;
    const { articleId } = match.params;
    const { article, comments } = singleArticle;
    const { userProfile } = user;
    const { profile } = userProfile;

    const { loader: isLoading } = isLoadingReducer;

    // sort comment based on the greater time in descending order
    const sortComment = comments.sort(
      (a, b) =>
        new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime()
    );
    const url = window.location.href;
    return (
      <React.Fragment>
        <NavBar />
        {isLoading && <Loader />}
        <div className="article-page">
          <Grid>
            <Grid.Row columns={3}>
              <Grid.Column computer={2} mobile={16}>
                <div className="section">
                  {article.author && (
                    <div className="social-sharing">
                      <div className="facebook-share">
                        <a
                          href={`https://www.facebook.com/dialog/feed?app_id=370972486842144
                          &redirect_uri=https://develop--hermes-ah.netlify.com
                          &link=${url}
                          &picture=${article.image_url}
                          &caption=${article.abstract}`}
                          target="_blank"
                        >
                          <i className="fab fa-facebook-f" />
                        </a>
                      </div>
                      <div className="twitter-share">
                        <a
                          href={`https://twitter.com/intent/tweet?text=${
                            article.title
                          } by ${article.author.first_name}${
                            article.author.last_name
                          } ${url}`}
                          target="_blank"
                          data-size="large"
                        >
                          <i className="fab fa-twitter" />
                        </a>
                      </div>
                      <div className="bookmark">
                        <i className="far fa-bookmark" />
                      </div>
                      <div className="like">
                        <i className="far fa-thumbs-up" />
                        <p>{article.likes_count}</p>
                      </div>
                    </div>
                  )}
                </div>
              </Grid.Column>
              <Grid.Column computer={10} mobile={16}>
                <Grid.Row>
                  <Grid.Column>
                    <div className="section">
                      {Object.keys(article).length ? (
                        <div>
                          <ReadingCard article={article} />
                          <Rate
                            articleId={articleId}
                            rateArticle={rateArticle}
                            likes={article.likes_count}
                            reportArticle={reportArticle}
                          />
                        </div>
                      ) : (
                        <p>Loading...</p>
                      )}
                    </div>
                  </Grid.Column>
                  <h3>Comments</h3>
                  {Object.keys(article).length && (
                    <InputComment
                      imageUrl={profile && profile.image_url}
                      articleId={article.id}
                      postComment={postComment}
                    />
                  )}
                  <div>
                    {sortComment.map(comment => (
                      <ViewComment
                        key={comment.id}
                        comment={comment}
                        imageUrl={article.author && article.author.image_url}
                        articleId={articleId}
                      />
                    ))}
                  </div>
                </Grid.Row>
              </Grid.Column>
              <Grid.Column computer={4} mobile={16}>
                <div className="section">Related article section</div>
              </Grid.Column>
            </Grid.Row>
          </Grid>
        </div>
      </React.Fragment>
    );
  }
}

ArticlePage.propTypes = {
  getSingleArticle: PropTypes.func.isRequired,
  postComment: PropTypes.func.isRequired,
  match: PropTypes.shape(PropTypes.objectOf).isRequired,
  singleArticle: PropTypes.shape({}).isRequired,
  rateArticle: PropTypes.func.isRequired,
  isLoadingReducer: PropTypes.shape({
    loader: PropTypes.bool,
  }).isRequired,
  reportArticle: PropTypes.func.isRequired,
  user: PropTypes.shape().isRequired,
};

export default ArticlePage;
