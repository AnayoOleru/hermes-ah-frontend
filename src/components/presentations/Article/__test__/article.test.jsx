import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ngFaker from 'ng-faker';
import ReadingCard from '../ReadingArticleCard';
import ArticlePage from '../ArticlePage';
import InputComment from '../InputComment';
import ViewComment from '../ViewComment';

const names = ngFaker.name.firstName();

const props = {
  article: {
    Keywords: [{ keywords: 'chemistry' }, { keywords: 'chemistry' }],
    author: { image_url: 'url', first_name: names, last_name: names },
    body: '<p>html</p>',
    image_url: 'url',
  },
  comments: [{ replies: [{ replier: {} }] }],
};
describe('Article', () => {
  it('reading card', () => {
    const wrapper = shallow(<ReadingCard {...props} />);
    expect(wrapper.find('div'));
  });
});

const middleware = [thunk];
const mockStore = configureMockStore(middleware);
const store = mockStore({});

const articleProps = {
  getSingleArticle: jest.fn(),
  postComment: jest.fn(),
  updateComment: jest.fn(),
  reset: jest.fn(),
  comment: { id: 'id', replies: [{ replier: { createdAt: 'time' } }] },
  match: { params: 'id', articleId: 'articleId' },
  singleArticle: props,
  isLoadingReducer: { loader: true },
  image_url: 'image',
  articleId: 'id',
  user: { userProfile: {} },
};

const inputCommentProps = {
  imageUrl: 'string',
  closeVal: 'string',
  articleId: 'string',
  btnValue: 'string',
  submitForm: jest.fn(),
  commentVal: 'string',
  handleChange: jest.fn(),
  enterKeyFormSubmit: jest.fn(),
  handleClose: jest.fn(),
};

const ArticlePageContainer = (
  <Provider store={store}>
    <ArticlePage
      match={articleProps.match}
      getSingleArticle={articleProps.getSingleArticle}
      singleArticle={articleProps.singleArticle}
      postComment={articleProps.postComment}
      isLoadingReducer={articleProps.isLoadingReducer}
      user={articleProps.user}
      reset={articleProps.reset}
      updateComment={articleProps.updateComment}
    />
  </Provider>
);

describe('Article Page', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(ArticlePageContainer);
    expect(wrapper).toMatchSnapshot();
  });

  it('reading card', () => {
    const wrapper = shallow(
      <ArticlePage
        match={articleProps.match}
        getSingleArticle={articleProps.getSingleArticle}
        singleArticle={articleProps.singleArticle}
        postComment={articleProps.postComment}
        isLoadingReducer={articleProps.isLoadingReducer}
        user={articleProps.user}
        reset={articleProps.reset}
        updateComment={articleProps.updateComment}
      />
    );
    expect(wrapper.find('div'));
  });

  it('InputComment component', () => {
    const wrapper = shallow(<InputComment {...inputCommentProps} />);
    expect(wrapper.find('div'));
  });

  it('should handle sortComment function', async () => {
    const wrapper = shallow(
      <ArticlePage
        match={articleProps.match}
        getSingleArticle={articleProps.getSingleArticle}
        singleArticle={articleProps.singleArticle}
        postComment={articleProps.postComment}
        isLoadingReducer={articleProps.isLoadingReducer}
        user={articleProps.user}
        reset={articleProps.reset}
        updateComment={articleProps.updateComment}
      />
    );
    const comment = [
      { id: 'id', replies: [{ replier: { createdAt: 'time' } }] },
    ];
    expect(await wrapper.instance().sortComment(comment));
  });

  const commentViewProps = {
    article: {
      id: 'id',
      author: { image_url: 'url', first_name: names, last_name: names },
    },
    comment: {
      id: 'id',
      replies: [{ replier: { createdAt: 'time' } }],
      commentOwner: { image_url: 'image' },
    },
    match: { params: 'id', articleId: 'articleId' },
    updateComment: jest.fn(),
  };

  it('shoud handle handleCommentInput function', () => {
    const wrapper = shallow(<ArticlePage {...articleProps} />);
    const event = { preventDefault: jest.fn(), target: { value: 'value' } };
    expect(wrapper.instance().handleCommentInput(event));
  });

  it('shoud handle sendComment function', () => {
    const wrapper = shallow(<ArticlePage {...articleProps} />);
    wrapper.setState({ commentVal: 'input' });
    const event = { preventDefault: jest.fn(), target: { id: 'id' } };
    expect(wrapper.instance().sendComment(event));
  });

  it('should handle onEnterSubmit function', () => {
    const wrapper = shallow(<ArticlePage {...articleProps} />);
    wrapper.setState({ commentVal: 'input' });
    const event = { preventDefault: jest.fn(), keyCode: 13, shiftKey: false };
    expect(wrapper.instance().onEnterSubmit(event));
  });

  it('should handle showCommentInput function', () => {
    const wrapper = shallow(<ViewComment {...commentViewProps} />);
    wrapper.setState({ input: true });
    const event = { preventDefault: jest.fn(), target: { id: 'id' } };
    expect(wrapper.instance().showCommentInput(event));
  });

  it('should handle toggleReplies function', () => {
    const wrapper = shallow(<ViewComment {...commentViewProps} />);
    wrapper.setState({ toggle: true });
    expect(wrapper.instance().toggleReplies());
  });

  it('should handle toggleCommentMenu function', () => {
    const wrapper = shallow(<ViewComment {...commentViewProps} />);
    wrapper.setState({ menu: true });
    expect(wrapper.instance().toggleCommentMenu());
  });

  it('should handle showEditComment function', () => {
    const wrapper = shallow(<ViewComment {...commentViewProps} />);
    wrapper.setState({ showEdit: true });
    expect(wrapper.instance().showEditComment());
  });

  it('should handle handleReplyInput function', () => {
    const wrapper = shallow(<ViewComment {...commentViewProps} />);
    wrapper.setState({ replyVal: 'value' });
    const event = { preventDefault: jest.fn(), target: { value: 'value' } };
    expect(wrapper.instance().handleReplyInput(event));
  });

  it('should handle editComment function', () => {
    const wrapper = shallow(<ViewComment {...commentViewProps} />);
    wrapper.setState({ replyVal: '' });
    const event = { preventDefault: jest.fn(), target: { id: 'id' } };
    expect(wrapper.instance().editComment(event));
  });
});
