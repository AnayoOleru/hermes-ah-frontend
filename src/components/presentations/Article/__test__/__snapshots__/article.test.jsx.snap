// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article Page should match snapshot 1`] = `ShallowWrapper {}`;
